### PR TITLE
Support options like charset and collation on DiscriminatedColumn

### LINKED ISSUE #10462: Add options (charset, collation) to DiscriminatorColumn to align it with Column options
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | no
| BC Break    | no

#### Summary

As stated in [related dbal issue](https://github.com/doctrine/dbal/pull/5869#issue-1556699509), when trying to declare charset and collation for discriminator column, the only option to use is `columnDefinition` (which causes the migration issue described in the dbal issue). I believe that [DiscriminatorColumn](https://github.com/doctrine/orm/blob/2.14.1/lib/Doctrine/ORM/Mapping/DiscriminatorColumn.php) is not that much different from regular [Column](https://github.com/doctrine/orm/blob/2.14.1/lib/Doctrine/ORM/Mapping/Column.php) and it should support the same features.

### PR DESCRIPTION
**Pull request by JanTvrdik**
Closes: https://github.com/doctrine/orm/issues/10462

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/Builder/ClassMetadataBuilder.php [hunk 1]
```diff
@@ -220,10 +220,11 @@ public function setSingleTableInheritance()
      * @param string $type
      * @param int    $length
      * @psalm-param class-string<BackedEnum>|null $enumType
+     * @psalm-param array<string, mixed> $options
      *
      * @return $this
      */
-    public function setDiscriminatorColumn($name, $type = 'string', $length = 255, ?string $columnDefinition = null, ?string $enumType = null)
+    public function setDiscriminatorColumn($name, $type = 'string', $length = 255, ?string $columnDefinition = null, ?string $enumType = null, array $options = [])
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/Builder/ClassMetadataBuilder.php:227]: Technically a breaking change for extending classes, but we didn't frown upon a similar change in https://github.com/doctrine/orm/pull/10288 , so maybe we should just make the class as final on 3.0.x.

**Review by greg0ire — COMMENTED** [review]
This looks like a new feature, and as such should target 2.15.x

File: lib/Doctrine/ORM/Mapping/Builder/ClassMetadataBuilder.php (hunk 1 above)
Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/Builder/ClassMetadataBuilder.php:227]: BTW, that PR modified docs/en/reference/php-mapping.rst , and this one should too I think. You might want to look at that PR for other changes that need to be done.

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 2]
```diff
@@ -3131,7 +3131,7 @@ public function addEntityListener($eventName, $class, $method)
      * @see getDiscriminatorColumn()
      *
      * @param mixed[]|null $columnDef
-     * @psalm-param DiscriminatorColumnMapping|array{name: string|null, fieldName?: string, type?: string, length?: int, columnDefinition?: string|null, enumType?: class-string<BackedEnum>|null}|null $columnDef
+     * @psalm-param DiscriminatorColumnMapping|array{name: string|null, fieldName?: string, type?: string, length?: int, columnDefinition?: string|null, enumType?: class-string<BackedEnum>|null, options?: array<string, mixed>}|null $columnDef
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: :thinking: I don't know why I left this `|array{…}` part in https://github.com/doctrine/orm/pull/10513 but it looks wrong now.

Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: https://github.com/doctrine/orm/pull/10602

Comment (greg0ire) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: Merged. please address the conflict.

File: tests/Doctrine/Tests/ORM/Mapping/AttributeReaderTest.php [hunk 3]
```diff
@@ -101,3 +122,21 @@ class TestTag
...
+#[Entity]
+#[InheritanceType('SINGLE_TABLE')]
+#[DiscriminatorColumn(name: 'discr', options: ['charset' => 'ascii', 'collation' => 'ascii_general_ci'])]
+#[DiscriminatorMap(['person' => TestPerson::class, 'employee' => TestEmployee::class])]
+class TestPerson
+{
+    #[Id]
+    #[Column(type: 'integer')]
+    #[GeneratedValue]
+    public int|null $id = null;
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Mapping/AttributeReaderTest.php]: While PHP 7.4 and below understands attributes as just comments, it cannot understand this union type.

**Review by greg0ire — APPROVED** [review]

**Review by derrabus — APPROVED** [review]

**Comment by greg0ire** [comment]
Thanks @JanTvrdik !

**Pull request merged**